test/system_benchmarks.cc

```diff
@@ -139,3 +142,117 @@ TEMPLATE_TEST_CASE_METHOD_SIG(
+
+void run_rsim_benchmark(size_t n_tris, size_t num_iter) {
+#ifndef NDEBUG
+	if(n_tris > 64 || num_iter > 50) { SKIP("Skipping larger-scale benchmark in debug build to save CI time"); }
+#endif
+
+	queue queue;
+
+	// we simply set kij to all 1s
+	const auto kij_size = range<2>(n_tris, n_tris);
+	std::vector<float> kij_data(n_tris * n_tris, 1.f);
+	buffer<float, 2> kij(kij_data.data(), kij_size);
+
+	// rad starts as all 0s, but we need a new buffer every time the benchmark starts
+	// otherwise we are not actually measuring the growing buffer access pattern
+	const auto rad_size = range<2>(num_iter, n_tris);
+	std::vector<float> rad_data(n_tris * num_iter, 0.f);
+
+	// this is a separate result buffer that survives the benchmark section to allow checking the correctness
+	buffer<float, 2> rad_result(rad_size);
+
+	BENCHMARK("iterations") {
+		buffer<float, 2> rad(rad_data.data(), rad_size);
+		// set the first line of rad to 1
+		queue.submit([&](handler& cgh) {
+			auto write_rad_mapper = [](chunk<1> chnk) -> subrange<2> {
+				return {
+				    id(0, chnk.offset.get(0)),  // offset
+				    range(1, chnk.range.get(0)) // range
+				};
+			};
+			accessor write_rad{rad, cgh, write_rad_mapper, write_only, no_init};
+			cgh.parallel_for(n_tris, [=](item<1> item) { write_rad[{0, item.get_id(0)}] = 1.f; });
+		});
+
+		for(size_t t = 1; t < num_iter; ++t) {
+			queue.submit([&](handler& cgh) {
+				// read everything written before the current timestep
+				auto read_rad_mapper = [t](chunk<2> chnk) -> subrange<2> {
+					return {
+					    id(0, 0),                         // offset
+					    range(t, chnk.global_size.get(0)) // range
+					};
+				};
+				// only need to write to radiosities of own triangles in current timestep
+				auto write_rad_mapper = [t](chunk<2> chnk) -> subrange<2> {
+					return {
+					    id(t, chnk.offset.get(0)),  // offset
+					    range(1, chnk.range.get(0)) // range
+					};
+				};
+
+				accessor read_kij{kij, cgh, access::one_to_one(), read_only};
+				accessor write_rad{rad, cgh, write_rad_mapper, write_only, no_init};
+				accessor read_rad{rad, cgh, read_rad_mapper, read_only};
+
+				cgh.parallel_for(kij_size, [=](item<2> item) {
+					float val = 0.f;
+					float included_items = 0.f;
+					for(size_t i = 0; i < t; ++i) {
+						// printf("i: %lu, t: %lu, id: %lu/%lu, read_rad: %f, read_kij: %f\n", i, t, item.get_id(0), item.get_id(1), read_rad[{i,
+						// item.get_id(0)}], read_kij[{item.get_id(0), item.get_id(1)}]);
+						val += read_rad[{i, item.get_id(0)}] * read_kij[{item.get_id(0), item.get_id(1)}];
+					}
+					val /= (float)t;
+					write_rad[{t, item.get_id(0)}] = val;
+				});
+			});
+		}
+		queue.submit([&](handler& cgh) {
+			accessor read_rad{rad, cgh, access::one_to_one(), read_only};
+			accessor write_rad_result{rad_result, cgh, access::one_to_one(), write_only};
+			cgh.parallel_for(rad_size, [=](item<2> item) { write_rad_result[item] = read_rad[item]; });
+		});
```

Comment:
Instead of doing this copy in each iteration, you could move the body of the `BENCHMARK` into a lambda that accepts a `bool verify` and call that once before running the actual benchmark.

Reply:
Do you think that's worth it? It's just one copy of a relatively small buffer that happens once after 50 / 500 iterations of the kernel.

Reply:
Discussed offline, and I was convinced this makes sense. Implemented.